src/domain_tests/TestRunner.cs

```diff
@@ -0,0 +1,1115 @@
+                    dynamic sys = Py.Import(""sys"");
+                    sys.path.append(new PyString(temp));
+                    dynamic test_mod = Py.Import(""domain_test_module.mod"");
+                    test_mod.{1}_reload();
+                }}
+                PythonEngine.Shutdown();
+            }}
+            catch (PythonException pe)
+            {{
+                throw new ArgumentException(message:pe.Message+""    ""+pe.StackTrace);
+            }}
+            catch (Exception e)
+            {{
+                Console.WriteLine(e.StackTrace);
+                throw;
+            }}
+            return 0;
+        }}
+    }}
+}}
+";
+        readonly static string PythonDllLocation = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Python.Runtime.dll");
+
+        static string TestPath = null;
+
+        public static int Main(string[] args)
+        {
+            TestCase testCase;
+            if (args.Length < 1)
+            {
+                testCase = Cases[0];
+            }
+            else
+            {
+                string testName = args[0];
+                Console.WriteLine($"-- Looking for domain reload test case {testName}");
+                testCase = Cases.First(c => c.Name == testName);
+            }
+
+            Console.WriteLine($"-- Running domain reload test case: {testCase.Name}");
+
+            SetupTestFolder(testCase.Name);
+
+            CreatePythonModule(testCase);
+            {
+                var runnerAssembly = CreateCaseRunnerAssembly(verb:"before");
+                CreateTestClassAssembly(testCase.DotNetBefore);
+                {
+                    var runnerDomain = CreateDomain("case runner before");
+                    RunAndUnload(runnerDomain, runnerAssembly);
+                }
+                {
+                    var runnerDomain = CreateDomain("case runner before (again)");
+                    RunAndUnload(runnerDomain, runnerAssembly);
+                }
+            }
+
+            {
+                var runnerAssembly = CreateCaseRunnerAssembly(verb:"after");
+                CreateTestClassAssembly(testCase.DotNetAfter);
+
+                // Do it twice for good measure
+                {
+                    var runnerDomain = CreateDomain("case runner after");
+                    RunAndUnload(runnerDomain, runnerAssembly);
+                }
+                {
+                    var runnerDomain = CreateDomain("case runner after (again)");
+                    RunAndUnload(runnerDomain, runnerAssembly);
+                }
+            }
+
+            // Don't delete unconditionally. It's sometimes useful to leave the
+            // folder behind to debug failing tests.
+            TeardownTestFolder();
+
+            return 0;
+        }
+
+        static void SetupTestFolder(string testCaseName)
+        {
+            TestPath = Path.Combine(Path.GetTempPath(), $"Python.TestRunner.{testCaseName}");
```

Comment:
I think some random number should also be added to indicate the overall run, so that different runs would not clash.

Reply:
They wouldn't, if the folder exists when setting up the test, it gets deleted first. Also, the executable can only run one test per invocation.

Reply:
I still see a problem with running tests for multiple Python versions on the same machine in parallel.

Reply:
Ah, I hadn't thought about that one. I'll add the python version and architecture to the folder name.

Reply:
I'll add the PID instead